src/System.Net.Http/src/System/Net/Http/Unix/CurlCallbacks.cs

```diff
@@ -0,0 +1,240 @@
+                    {
+                        state.ResponseMessage.SignalHeadersReceived();
+                        return totalSize;
+                    }
+
+                    // TODO: Understand scenarios where multiple sets of headers are received
+                    if (!TryParseStatusLine(response, responseHeader))
+                    {
+                        int colonIndex = responseHeader.IndexOf(':');
+
+                        // Skip malformed header lines that are missing the colon character.
+                        if (colonIndex > 0)
+                        {
+                            string headerName = responseHeader.Substring(0, colonIndex);
+                            string headerValue = responseHeader.Substring(colonIndex + 1);
+
+                            if (!response.Headers.TryAddWithoutValidation(headerName, headerValue))
+                            {
+                                // TODO: Skip compression related headers
+                                response.Content.Headers.TryAddWithoutValidation(headerName, headerValue);
+                            }
+                        }
+                    }
+                    return totalSize;
+                });
+        }
+
+        private ulong CurlReceiveBodyCallback(
+            IntPtr buffer,
+            ulong size,
+            ulong nitems,
+            IntPtr context)
+        {
+            return ExecuteCallback(buffer, size, nitems, context,
+                (ptr, totalSize, state) =>
+                {
+#if DEBUG
+                    LogCurlBuffer(ptr, totalSize);
+#endif
+
+                    // TODO: Handle compression and chunked-mode
+
+                    var byteBuffer = new byte[totalSize];
+                    Marshal.Copy(ptr, byteBuffer, 0, byteBuffer.Length);
+                    Stream contentStream = state.ResponseMessage.ContentStream;
+                    contentStream.Write(byteBuffer, 0, byteBuffer.Length);  // will block if no readers
+                    contentStream.Flush();
+                    return totalSize;
+                });
+        }
+
+
+        private ulong CurlSendCallback(
+            IntPtr buffer,
+            ulong size,
+            ulong nitems,
+            IntPtr context)
+        {
+            return ExecuteCallback(buffer, size, nitems, context,
+                (ptr, totalSize, state) =>
+                {
+                    state.CancellationToken.ThrowIfCancellationRequested();
+                    Stream contentStream = state.RequestContentStream;
+
+                    // TODO: Handle chunked-mode
+
+                    int bytesRead = 0;
+                    while (bytesRead < (int)totalSize)
+                    {
+                        state.CancellationToken.ThrowIfCancellationRequested();
+                        var byteBuffer = new byte[totalSize];
+                        int numBytes = contentStream.Read(byteBuffer, bytesRead, (int)totalSize - bytesRead);
```

Comment:
Do we know that all of the data in the contentStream is available promptly, or could this call to Read block waiting for more data to arrive?  If the latter, shouldn't we be implementing this asynchronously, using ReadAsync?